nixos/test-driver: improve error reporting and assertions

Ma27 opened the pull request:
This essentially does the following things (and I'd recommend to review commit-by-commit):

* add `t` as instance of `unittest.TestCase` to write assertions in the style of `t.assertEqual` etc.. This has the benefit that we get proper reporting for assertion failures for free in contrast to an empty `AssertionError`.

  I considered using pytest before, but this requires messing around with its internals and probably even touching the `ast` and we already came to the conclusion in #345948 that we don't really want to have this in our framework.

  I'm considering to create a feature-request to pytest for a reasonable entry-point, but I'm not too optimistic about this.

  The name `t` is chosen as shorthand since the assertion methods are already kinda verbose and having something like `tester.assertEqual()` feels quite annoying to type. Happy to discuss other names, though ;) 

* Create a new exception class-structure that - when caught - only prints stack frames in the `testScript` (this indicates a test failure, not a bug in the framework, so the frames from the framework are just visual noise) and prefixes the error with `!!! ` (in red) so that it's easier to spot what the interesting lines are between the other logs.

* Switched to the `ipython` integration of `ptpython` for interaction: this handles `sys.exit` properly and thus also fixes #180089 (and doesn't cause an exit on failing `machine.succeed` on this branch).

* Added a `machine: ` prefix to the log message indicating a finished subtest.

Thank you @bew for working on this together!

Example:
![2025-03-18_16h48m47s_screenshot](https://github.com/user-attachments/assets/76440791-8c2a-4654-8bcf-7a9e0e610efc)

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Ma27 committed:
nixos/test-driver: integrate Python `unittest` assertions

Replaces / Closes #345948

I tried to integrate `pytest` assertions because I like the reporting,
but I only managed to get the very basic thing and even that was messing
around a lot with its internals.

The approach in #345948 shifts too much maintenance effort to us, so
it's not really desirable either.

After discussing with Benoit on Ocean Sprint about this, we decided that
it's probably the best compromise to integrate `unittest`: it also
provides good diffs when needed, but the downside is that existing tests
don't benefit from it.

This patch essentially does the following things:

* Add a new global `t` that is an instance of a `unittest.TestCase`
  class. I decided to just go for `t` given that e.g.
  `tester.assertEqual` (or any other longer name) seems quite verbose.

* Use a special class for errors that get special treatment:
  * The traceback is minimized to only include frames from the
    testScript: in this case I don't really care about anything else and
    IMHO that's just visual noise.

    This is not the case for other exceptions since these may indicate a
    bug and then people should be able to send the full traceback to the
    maintainers.
  * Display the error, but with `!!!` as prefix to make sure it's
    easier to spot in between other logs.

This looks e.g. like

    !!! Traceback (most recent call last):
    !!!   File "<string>", line 7, in <module>
    !!!     foo()
    !!!   File "<string>", line 5, in foo
    !!!     t.assertEqual({"foo":[1,2,{"foo":"bar"}]},{"foo":[1,2,{"bar":"foo"}],"bar":[1,2,3,4,"foo"]})
    !!!
    !!! NixOSAssertionError: {'foo': [1, 2, {'foo': 'bar'}]} != {'foo': [1, 2, {'bar': 'foo'}], 'bar': [1, 2, 3, 4, 'foo']}
    !!! - {'foo': [1, 2, {'foo': 'bar'}]}
    !!! + {'bar': [1, 2, 3, 4, 'foo'], 'foo': [1, 2, {'bar': 'foo'}]}
    cleanup
    kill machine (pid 9)
    qemu-system-x86_64: terminating on signal 15 from pid 6 (/nix/store/wz0j2zi02rvnjiz37nn28h3gfdq61svz-python3-3.12.9/bin/python3.12)
    kill vlan (pid 7)
    (finished: cleanup, in 0.00 seconds)

Co-authored-by: bew <bew@users.noreply.github.com>

Ma27 committed:
nixos/test-driver: use RequestedAssertionFailed/TestScriptError in Machine class

I think it's reasonable to also have this kind of visual distinction
here between test failures and actual errors from the test framework.

A failing `machine.require_unit_state` now lookgs like this for
instance:

    !!! Traceback (most recent call last):
    !!!   File "<string>", line 3, in <module>
    !!!     machine.require_unit_state("postgresql","active")
    !!!
    !!! RequestedAssertionFailed: Expected unit 'postgresql' to to be in state 'active' but it is in state 'inactive'

Co-authored-by: Benoit de Chezelles <bew@users.noreply.github.com>

Ma27 commented on nixos/lib/test-driver/src/test_driver/logger.py:
```diff
@@ -177,6 +188,8 @@ def __init__(self) -> None:
         if "machine" in attributes:
             return f"{attributes['machine']}: {message}"
+        if "prefix" in attributes:
+            return f"{attributes['prefix']}{message}"
```
I think this was from a previous iteration and isn't needed anymore.

tfc commented on nixos/lib/test-driver/src/test_driver/errors.py:
```diff
@@ -0,0 +1,13 @@
+from unittest import TestCase
+
```
can you please add comments/docstrings to describe what these errors semantically mean? I don't understand what a NixOS error shall be

Ma27 replied:
Done.
And I agree, `NixOSError` isn't too descriptive, changed :)

tfc commented on nixos/lib/test-driver/src/test_driver/errors.py:
```diff
@@ -0,0 +1,16 @@
+
+class NixOSRequestedAssertionFailed(TestScriptError):
+    """
+    Exception raised when a requested assertion,
+    e.g. `machine.succeed()` or `t.assertEqual()`, fails.
```
i appreciate the effort to make the exception names more descriptive, but i still don't understand the difference between both. I will btw. join you guys on the oceansprint tomorrow at dinner and the rest of the week - maybe we discuss the changes in person?

bew replied:
The idea for the difference is to separate an assertion that you (the user) explicitly requested to check in the test script (where you have an explicit error type before the whole message), vs a transitive error in the test driver for some reason that you didn't necessarily expect to happen 🤔

K900 commented:
Generally big fan of the direction, will probably not have spoons to review in detail.

bew commented on nixos/lib/test-driver/src/test_driver/errors.py:
```diff
@@ -0,0 +1,16 @@
+    """
+
+
+class NixOSRequestedAssertionFailed(TestScriptError):
```
About naming, I'm thinking the `NixOS` isn't really useful 🤔

Ma27 committed:
nixos/test-driver: log associated machine for `self.nested`

When doing `machine.succeed(...)` or something similar, it's now clear
that the command `...` was issued on `machine`.

Essentially, this results in the following diff in the log:

    -(finished: waiting for unit default.target, in 13.47 seconds)
    +machine: (finished: waiting for unit default.target, in 13.47 seconds)
    (finished: subtest: foobar text lorem ipsum, in 13.47 seconds)

bew reviewed (commented):
tried a different approach for the docstrings, by only mentioning in the 2nd error what's specific about it and why it exists.

Ma27 committed:
nixos/test-driver: restructure error classes

After a discussion with tfc, we agreed that we need a distinction
between errors where the user isn't at fault (e.g. OCR failing - now
called `MachineError`) and errors where the test actually failed (now
called `RequestedAssertionFailed`).

Both get special treatment from the error handler, i.e. a `!!!` prefix
to make it easier to spot visually.

However, only `RequestedAssertionFailed` gets the shortening of the
traceback, `MachineError` exceptions may be something to report and
maintainers usually want to see the full trace.

Suggested-by: Jacek Galowicz <jacek@galowicz.de>

Ma27 committed:
nixos/test-driver: use ipython via ptpython

Closes #180089

I realized that the previous commits relying on `sys.exit` for dealing
with `MachineError`/`RequestedAssertionFailed` exit the interactive
session which is kinda bad.

This patch uses the ipython driver: it seems to have equivalent features
such as auto-completion and doesn't stop on SystemExit being raised.

This also fixes other places where this happened such as things calling
`log.error` on the CompositeLogger.

Ma27 commented:
@tfc as discussed yesterday, there are now two classes:
* `MachineError` that gets the full stacktrace, but a special error-prefix
* `RequestedAssertionFailed` for failing `unittest` assertions and failing `machine.succeed` etc. with a prefix as well and a collapsed stack trace.

I pushed another commit switching to the `ipython` integration of ptpython to not stop interactive sessions on `sys.exit` (fixes #180089 and also the handlers on this branch that use sys.exit after the exception printing).

Let me know if I missed something!

bew commented on nixos/lib/test-driver/src/test_driver/errors.py:
```diff
@@ -0,0 +1,20 @@
+    e.g. a failing `t.assertEqual(...)` or `machine.succeed(...)`.
+
+    This gets special treatment in error reporting: i.e. it gets
+    `!!!` as prefix just as `MachineError`, but all stack frames that are
+    not from `testScript` also get removed.
```
```suggestion
    not from `testScript` get removed.
```

bew replied:
Or what about removing the negative? As in:
```suggestion
    `!!!` as prefix just as `MachineError`, but only stack frames coming
    from `testScript` will show up in logs.
```

bew commented on nixos/lib/test-driver/src/test_driver/driver.py:
```diff
@@ -163,7 +180,34 @@ def test_script(self) -> None:
+                sys.exit(1)
+            except RequestedAssertionFailed:
+                exc_type, exc, tb = sys.exc_info()
+                filtered = [
```
```suggestion
                # Keep only stack frames mentioning the test script
                # (note: because the script is not a real file, the frame filename is `<string>`)
                filtered = [
```

bew reviewed (approved):
Just a few nitpicks and a comment to explain what's going on in that exception handler.
Good call on the `MachineError` base exception class, naming is hard 👍

Ma27 committed:
nixos/test-driver: improve wording on comments about new error handling

Co-authored-by: Benoit de Chezelles <bew@users.noreply.github.com>

ctheune reviewed (approved):
Thanks! I'm a bit sad that we couldn't go the pytest-style route, but hey, this is a big improvement, thanks!

Ma27 commented:
@tfc would you mind taking another look? Would be kinda interested in having the non-controversial part in 25.05 at least :grin:

tfc commented:
> @tfc would you mind taking another look? Would be kinda interested in having the non-controversial part in 25.05 at least 😁

Yes, absolutely! I was quite a bit swamped in the last weeks. I hope to go through this in the easter holidays.

tfc commented:
@ofborg test nat.firewall networking.scripted.link installer.simpleProvided installer.separateBootFat networking.scripted.virtual printing keymap.azerty keymap.dvorak-programmer boot-stage1 installer.swraid keymap.neo nfs4.simple i3wm udisks2 networking.networkd.loopback containers-ip ecryptfs login installer.simpleLabels php.httpd zfs.installer predictable-interface-names.unpredictable predictable-interface-names.unpredictableNetworkd mutableUsers

tfc commented:
@ofborg test nat.firewall-conntrack networking.scripted.sit networking.networkd.virtual networking.scripted.dhcpSimple firewall firefox networking.scripted.macvlan proxy firefox-esr sddm.default boot.uefiCdrom networking.scripted.static networking.scripted.routes installer.lvm networking.networkd.vlan containers-imperative boot.biosUsb xfce networking.networkd.dhcpOneIf installer.simpleUefiSystemdBoot keymap.qwertz gnome-xorg shadow networking.networkd.macvlan networking.networkd.routes

tfc commented:
i would generally merge this, but i wanted to see a variety of tests to run through via ofborg so we don't forget anything. however, i wonder why these tests don't really run through...

Ma27 commented:
Thanks!

Regarding the Darwin failures: https://github.com/NixOS/nixpkgs/pull/390996/checks?check_run_id=40837905334
I don't know the state of the Darwin fixes for the test driver, but this looks unrelated?

Regarding aarch64-linux: https://logs.ofborg.org/?key=nixos%2Fnixpkgs.390996&attempt_id=307c4594-4911-4c81-973c-c582172f2ad3
E.g. `client1` from the networkd/macvlan test complains from early on about the backdoor being not available while waiting for the VM to finish booting. GIven the amount of tests failing because of that on aarch64-linux and the fact that I cannot reproduce, I'm tempted to consider this a transient problem (probably caused by too many tests running at the same time on a builder?).

wdyt?

Ma27 commented:
cc @tfc I see no obvious regressions or did I miss anything? Would you be OK with merging this then? :)

tfc commented:
@Ma27 thank you for the good work!
